Fixed #138 & lots of refactors and cleaning

MichaelHillcox opened the pull request:
So this is a big one that I want to keep working on but for now this is a good start. 

- I've cleaned up the packages quite a lot and changed them to follow the recommeded layout from forge. 
- I've renamed some classes to follow a logical structure
- I've extended GenricGadget to supply more common functionality to the other Gadgets.
- Fixed #138 
- Removed some Depercated calls and replaced them with new ones 😄 
- Added a Creative Tab for the mod.

I've not changed any functionality specifically so logically nothing should be broken... he says... other than giving `GenricGadget` 5 new methods `getGadget` `getEnergyCost` `getDamagePerUse` `canUse` `applyDamage`. All pretty self explantary but the idea is that they all use a single funtion to applyDamage ( reduce engery if enabled ). And if the item can be used, has power / durability. If you are in creative then you can always use the tools. These changes can be seen here https://github.com/Direwolf20-MC/BuildingGadgets/pull/139/commits/22de097d3e7f852a2e7bb1a834e7eec60abcfac5#diff-229267ce220b32a7c48b90dfb9047194 and https://github.com/Direwolf20-MC/BuildingGadgets/blob/22de097d3e7f852a2e7bb1a834e7eec60abcfac5/src/main/java/com/direwolf20/buildinggadgets/items/GenericGadget.java#L125

If you would like a single pr for the #138 fix then let me know.

MajorTuvok commented on src/main/java/com/direwolf20/buildinggadgets/common/network/PacketAnchorKey.java:
```diff
@@ -34,20 +36,20 @@ public IMessage onMessage(PacketAnchorKey message, MessageContext ctx) {
             EntityPlayerMP playerEntity = ctx.getServerHandler().player;
             ItemStack heldItem = playerEntity.getHeldItem(EnumHand.MAIN_HAND);
-            if (!(heldItem.getItem() instanceof GenericGadget)) {
+            if (!(heldItem.getItem() instanceof GadgetGeneric)) {
```
A case for the static Method GadgetGeneric.getGadget

MajorTuvok commented on src/main/java/com/direwolf20/buildinggadgets/common/network/PacketChangeRange.java:
```diff
@@ -33,24 +35,24 @@ public IMessage onMessage(PacketChangeRange message, MessageContext ctx) {
             EntityPlayerMP playerEntity = ctx.getServerHandler().player;
             ItemStack heldItem = playerEntity.getHeldItem(EnumHand.MAIN_HAND);
-            if (!(heldItem.getItem() instanceof GenericGadget)) {
+            if (!(heldItem.getItem() instanceof GadgetGeneric)) {
```
A case for the static Method GadgetGeneric.getGadget

MajorTuvok commented on src/main/java/com/direwolf20/buildinggadgets/common/network/PacketCopyCoords.java:
```diff
@@ -51,23 +51,23 @@ public IMessage onMessage(PacketCopyCoords message, MessageContext ctx) {
             EntityPlayerMP playerEntity = ctx.getServerHandler().player;
             ItemStack heldItem = playerEntity.getHeldItem(EnumHand.MAIN_HAND);
-            if (!(heldItem.getItem() instanceof CopyPasteTool)) {
+            if (!(heldItem.getItem() instanceof GadgetCopyPaste)) {
```
A case for the static Method GadgetCopyPaste.getGadget (requires override)

MajorTuvok replied:
Replace with getGadget.
Please add getGadget override (GadgetCopyPaste)...

MajorTuvok commented on src/main/java/com/direwolf20/buildinggadgets/common/network/PacketDestructionGUI.java:
```diff
@@ -54,17 +54,17 @@ public IMessage onMessage(PacketDestructionGUI message, MessageContext ctx) {
             EntityPlayerMP playerEntity = ctx.getServerHandler().player;
             ItemStack heldItem = playerEntity.getHeldItem(EnumHand.MAIN_HAND);
-            if (!(heldItem.getItem() instanceof DestructionTool)) {
+            if (!(heldItem.getItem() instanceof GadgetDestruction)) {
```
A case for the static Method GadgetDestruction.getGadget (requires override)

MajorTuvok replied:
Replace with getGadget.
Please add getGadget override (GadgetDestruction)...

MajorTuvok commented on src/main/java/com/direwolf20/buildinggadgets/common/network/PacketPasteGUI.java:
```diff
@@ -48,15 +48,15 @@ public IMessage onMessage(PacketPasteGUI message, MessageContext ctx) {
             EntityPlayerMP playerEntity = ctx.getServerHandler().player;
             ItemStack heldItem = playerEntity.getHeldItem(EnumHand.MAIN_HAND);
-            if (!(heldItem.getItem() instanceof CopyPasteTool)) {
+            if (!(heldItem.getItem() instanceof GadgetCopyPaste)) {
```
A case for the static Method GadgetCopyPaste.getGadget (requires override)

MajorTuvok replied:
Replace with getGadget.
Please add getGadget override (GadgetCopyPaste)...

MajorTuvok commented on src/main/java/com/direwolf20/buildinggadgets/common/network/PacketToggleMode.java:
```diff
@@ -44,21 +44,21 @@ public IMessage onMessage(PacketToggleMode message, MessageContext ctx) {
             EntityPlayerMP playerEntity = ctx.getServerHandler().player;
             ItemStack heldItem = playerEntity.getHeldItem(EnumHand.MAIN_HAND);
-            if (!(heldItem.getItem() instanceof GenericGadget)) {
+            if (!(heldItem.getItem() instanceof GadgetGeneric)) {
```
A case for the static Method GadgetGeneric.getGadget

MajorTuvok commented on src/main/java/com/direwolf20/buildinggadgets/common/network/PacketToggleMode.java:
```diff
@@ -44,21 +44,21 @@ public IMessage onMessage(PacketToggleMode message, MessageContext ctx) {
-            } else if (!heldItem.isEmpty() && heldItem.getItem() instanceof CopyPasteTool) {
-                CopyPasteTool copyPasteTool = (CopyPasteTool) (heldItem.getItem());
-                copyPasteTool.setMode(playerEntity, heldItem, message.mode);
+            if (!heldItem.isEmpty() && heldItem.getItem() instanceof GadgetBuilding) {
```
Maybe move this heldItem.isEmpty check to an enclosing if, as to avoid verbose multiple isEmpty checks?

MajorTuvok commented on src/main/java/com/direwolf20/buildinggadgets/common/network/PacketUndoKey.java:
```diff
@@ -33,24 +35,24 @@ public IMessage onMessage(PacketUndoKey message, MessageContext ctx) {
             EntityPlayerMP playerEntity = ctx.getServerHandler().player;
             ItemStack heldItem = playerEntity.getHeldItem(EnumHand.MAIN_HAND);
-            if (!(heldItem.getItem() instanceof GenericGadget)) {
+            if (!(heldItem.getItem() instanceof GadgetGeneric)) {
```
A case for the static Method GadgetGeneric.getGadget

MajorTuvok commented on src/main/java/com/direwolf20/buildinggadgets/common/tools/ToolRenders.java:
```diff
@@ -55,9 +59,9 @@ public static void renderBuilderOverlay(RenderWorldLastEvent evt, EntityPlayer p
             if (startBlock != ModBlocks.effectBlock.getDefaultState()) {
                 ItemStack heldItem = player.getHeldItemMainhand(); //Get the item stack and the block that we'll be rendering (From the Itemstack's NBT)
-                if (!(heldItem.getItem() instanceof BuildingTool)) {
+                if (!(heldItem.getItem() instanceof GadgetBuilding)) {
```
A case for the static Method GadgetBuilding.getGadget (requires override)

MajorTuvok replied:
Replace with getGadget.
Please add getGadget override (GadgetBuilding)...

MajorTuvok commented on src/main/java/com/direwolf20/buildinggadgets/common/tools/ToolRenders.java:
```diff
@@ -187,9 +191,9 @@ public static void renderExchangerOverlay(RenderWorldLastEvent evt, EntityPlayer
             if (startBlock != ModBlocks.effectBlock.getDefaultState()) {
                 ItemStack heldItem = player.getHeldItemMainhand(); //Get the item stack and the block that we'll be rendering (From the Itemstack's NBT)
-                if (!(heldItem.getItem() instanceof ExchangerTool)) {
+                if (!(heldItem.getItem() instanceof GadgetExchanger)) {
```
A case for the static Method GadgetExchanger.getGadget (requires override)

MajorTuvok commented on src/main/java/com/direwolf20/buildinggadgets/common/tools/ToolRenders.java:
```diff
@@ -314,24 +318,24 @@ public static void renderExchangerOverlay(RenderWorldLastEvent evt, EntityPlayer
 
         ItemStack heldItem = player.getHeldItemMainhand(); //Get the item stack and the block that we'll be rendering (From the Itemstack's NBT)
-        if (!(heldItem.getItem() instanceof DestructionTool)) {
+        if (!(heldItem.getItem() instanceof GadgetDestruction)) {
```
A case for the static Method GadgetDestruction.getGadget (requires override)

MajorTuvok commented on src/main/java/com/direwolf20/buildinggadgets/common/items/gadgets/GadgetGeneric.java:
```diff
@@ -96,4 +102,48 @@ public boolean getIsRepairable(ItemStack toRepair, ItemStack repair) {
+        return heldItem;
+    }
+
+    public int getEnergyCost() {
```
Config? or is this always overriden?

MajorTuvok commented on src/main/java/com/direwolf20/buildinggadgets/common/items/gadgets/GadgetGeneric.java:
```diff
@@ -96,4 +102,48 @@ public boolean getIsRepairable(ItemStack toRepair, ItemStack repair) {
+        return 200;
+    }
+
+    public int getDamagePerUse() {
```
Config? or is this always overriden?

MichaelHillcox commented on src/main/java/com/direwolf20/buildinggadgets/common/items/gadgets/GadgetBuilding.java:
```diff
@@ -109,15 +118,17 @@ public void addInformation(ItemStack stack, @Nullable World world, List<String>
     public EnumActionResult onItemUse(EntityPlayer player, World world, BlockPos pos, EnumHand hand, EnumFacing side, float hitX, float hitY, float hitZ) {
         //On item use, if sneaking, select the block clicked on, else build -- This is called when a block in clicked on
+        if (!world.isRemote)
+            return EnumActionResult.FAIL;
```
You're right. I've inverted the logic here. I was turning it into a return early statement but looks like I've forgot to remove the !

MichaelHillcox commented on src/main/java/com/direwolf20/buildinggadgets/common/items/gadgets/GadgetBuilding.java:
```diff
@@ -109,15 +118,17 @@ public void addInformation(ItemStack stack, @Nullable World world, List<String>
     @Override
     public EnumActionResult onItemUse(EntityPlayer player, World world, BlockPos pos, EnumHand hand, EnumFacing side, float hitX, float hitY, float hitZ) {
         //On item use, if sneaking, select the block clicked on, else build -- This is called when a block in clicked on
+        if (!world.isRemote)
```
```suggestion
        if (world.isRemote)
```

MajorTuvok replied:
Oh, I actually missed that... XD
Thought what I meant is that in the original implementation always returned SUCCESS, though now there is a case where it returns FAIL - but I guess that returning Failure is the way to go...

MichaelHillcox commented on src/main/java/com/direwolf20/buildinggadgets/common/items/gadgets/GadgetDestruction.java:
```diff
@@ -62,6 +59,16 @@ public int getEnergyMax() {
+    }
+
+    @Override
+    public int getDamagePerUse() {
```
see https://github.com/Direwolf20-MC/BuildingGadgets/pull/139#discussion_r233977984

MajorTuvok commented on src/main/java/com/direwolf20/buildinggadgets/common/items/gadgets/GadgetBuilding.java:
```diff
@@ -220,14 +229,11 @@ private boolean build(EntityPlayer player, ItemStack stack) {
         return true;
     }
 
```
Suggestion for override implementation...
```suggestion
public static GadgetBuilding getGadget(EntityPlayer p) {
  ItemStack heldItem = player.getHeldItemMainhand();
  if (!(heldItem.getItem() instanceof GadgetBuilding)) {
     heldItem = player.getHeldItemOffhand();
     if (!(heldItem.getItem() instanceof GadgetBuilding)) {
        return null;
     }
  }
  return heldItem;
}
```

MichaelHillcox replied:
see https://github.com/Direwolf20-MC/BuildingGadgets/pull/139/commits/85b80b12a0210d16305a0750afc6b2c2fd873a6a

MajorTuvok commented on src/main/java/com/direwolf20/buildinggadgets/common/items/gadgets/GadgetCopyPaste.java:
```diff
@@ -499,55 +508,40 @@ private static boolean findBlocks(World world, BlockPos start, BlockPos end, Ite
+        if (pos.getY() < 0 || state.equals(Blocks.AIR.getDefaultState()) || !player.isAllowEdit())
+            return;
+
+        ItemStack heldItem = getGadget(player);
```
Please add getGadget override (GadgetCopyPaste)...

MajorTuvok commented on src/main/java/com/direwolf20/buildinggadgets/common/items/gadgets/GadgetDestruction.java:
```diff
@@ -306,14 +313,16 @@ public static boolean validBlock(World world, BlockPos voidPos, EntityPlayer pla
+        if (currentBlock.getBlockHardness(world, voidPos) < 0) return false;
+
+        ItemStack tool = player.getHeldItemMainhand(); //Get the item stack and the block that we'll be rendering (From the Itemstack's NBT)
+        if (!(tool.getItem() instanceof GadgetDestruction)) {
```
Replace with getGadget.
Please add getGadget override (GadgetDestruction)...

MajorTuvok commented on src/main/java/com/direwolf20/buildinggadgets/common/items/gadgets/GadgetExchanger.java:
```diff
@@ -220,13 +220,11 @@ private boolean exchange(EntityPlayer player, ItemStack stack) {
-            }
-        }
+
+        ItemStack heldItem = getGadget(player);
```
Please add getGadget override (GadgetExchanger)...

MajorTuvok commented on src/main/java/com/direwolf20/buildinggadgets/common/items/gadgets/GadgetExchanger.java:
```diff
@@ -261,13 +259,11 @@ private boolean exchangeBlock(World world, EntityPlayer player, BlockPos pos, IB
-            }
-        }
+
+        ItemStack tool = getGadget(player);
```
Please add getGadget override (GadgetExchanger)...

MajorTuvok commented on src/main/java/com/direwolf20/buildinggadgets/common/tools/ToolRenders.java:
```diff
@@ -187,9 +191,9 @@ public static void renderExchangerOverlay(RenderWorldLastEvent evt, EntityPlayer
             if (startBlock != ModBlocks.effectBlock.getDefaultState()) {
                 ItemStack heldItem = player.getHeldItemMainhand(); //Get the item stack and the block that we'll be rendering (From the Itemstack's NBT)
-                if (!(heldItem.getItem() instanceof ExchangerTool)) {
+                if (!(heldItem.getItem() instanceof GadgetExchanger)) {
```
Replace with GadgetExchanger.getGadget
Please add getGadget override (GadgetExchanger)...

MichaelHillcox replied:
see https://github.com/Direwolf20-MC/BuildingGadgets/pull/139/commits/85b80b12a0210d16305a0750afc6b2c2fd873a6a

MajorTuvok commented on src/main/java/com/direwolf20/buildinggadgets/common/tools/ToolRenders.java:
```diff
@@ -314,24 +318,24 @@ public static void renderExchangerOverlay(RenderWorldLastEvent evt, EntityPlayer
 
         ItemStack heldItem = player.getHeldItemMainhand(); //Get the item stack and the block that we'll be rendering (From the Itemstack's NBT)
-        if (!(heldItem.getItem() instanceof DestructionTool)) {
+        if (!(heldItem.getItem() instanceof GadgetDestruction)) {
```
Replace with GadgetDestruction.getGadget
Please add getGadget override (GadgetDestruction)...

MichaelHillcox replied:
see https://github.com/Direwolf20-MC/BuildingGadgets/pull/139/commits/85b80b12a0210d16305a0750afc6b2c2fd873a6a

MichaelHillcox commented on src/main/java/com/direwolf20/buildinggadgets/common/items/gadgets/GadgetBuilding.java:
```diff
@@ -185,13 +196,11 @@ private boolean build(EntityPlayer player, ItemStack stack) {
-            }
-        }
+
+        ItemStack heldItem = getGadget(player);
```
Fixed 👍 see https://github.com/Direwolf20-MC/BuildingGadgets/pull/139/commits/85b80b12a0210d16305a0750afc6b2c2fd873a6a

MichaelHillcox commented:
Thats all fixed. I've noticed a few more deprecated calls in one of the Block classes so I'll sort them out and start work on cleaning up the GUI a bit but I wanna get this pull in @Direwolf20-MC before I continue with Gui and other non codebase changes.

Direwolf20-MC commented:
Wow thanks guys that’s a lotta changes! Unfortunately I’m out of town for a while so won’t get to reviewing it before the end of the month :(

MajorTuvok commented:
No Problem @MichaelHillcox...
Thank you for submitting and implementing the PR...

MajorTuvok commented:
@MichaelHillcox I just noticed a Throwable catch Block in DestructionGui...
Could you change that to an NumberFormatException?

Direwolf20-MC commented:
Hey guys, just wanted to touch base on this. I honestly don't have time to review all these code changes, its a LOT and my time constraints are the main issue. 

1) How confident are you that there are no breaking changes here?
2) Are you done (at least for now) such that if I accept the merge it'll work just fine?

Thanks!

@MichaelHillcox @MajorTuvok

MichaelHillcox commented:
When I started the refactors I focused on not changing any functionality. The majority of the changes are files moving to clean up packages and files being renamed. The other changes which me and @MajorTuvok have been working on have been regarding minimizing code duplication. 

I've also been working on other change in a different branch but based on these changes and I'm yet to have ran into an issue and I've been abusing the mod quite a bit so hopefully it should just work fine. 

@Direwolf20-MC I plan on staying around to keep improving the mod so if you wanted to assign me as a Contributor to the repo then I'd happily keep up with the issues and the etc. Just an idea :P Would be happy to talk more about such a thing if you wanted to :)

MichaelHillcox commented:
Haha, i totally get that :P 

To make it simple to understand what has changed:
Packages:
- Introduced to main packages `client`, `common`. Client is for client specific things, common is for everything else
- Introducted a package within items called `gadgets` this contains all the gadgets
   - With this the names of all the tools have been changed to prefixs of `Gadget...` (`GadgetDestruction` for example)
- Introduced a `client.events` package to keep all of the event specific classes in one place

Code Changes:
- Introduced a `handleEventInput` method to remove code dupe seen here https://github.com/Direwolf20-MC/BuildingGadgets/pull/139/files#diff-eb50b6ba37c3167c741c14b36cf8ad04
- Solved #141 seen here https://github.com/Direwolf20-MC/BuildingGadgets/pull/139/files#diff-873b65b0463fe4a2eb68f48fe92acddc
- Updated and line like `ResourceLocation("buildinggadgets:textures/ui/buildtome.png")` to `new ResourceLocation(BuildingGadgets.MODID,"textures/ui/buildtome.png")` to prevent issues with future forge versions.

Main additions:
- Expanded `GenricGadget` or what is now `GadgetGeneric` to have 5 new methods `getGadget` `getEnergyCost` `getDamagePerUse` `canUse` `applyDamage`
  - Apply Damage applies the correct damage type either durability or power
  - Can use checks if the item can be used and this will take in the `getEnergyCost` method that is overwritten in each gadget class.
  - getGadget is also overwritten in each gadget class to allo wus to changes lines like this https://i.imgur.com/jvKQ80l.png to this https://i.imgur.com/84xo21m.png
  - All these changes can be seen here https://github.com/Direwolf20-MC/BuildingGadgets/pull/139/files#diff-280c90826882a5a32ddb18c21397f0fc
- Added a creative tab for all the Building Tools items, blocks, etc.

I also cleaned the lang file up https://github.com/Direwolf20-MC/BuildingGadgets/pull/139/files#diff-22d99c90d5e172d2f64cb3a04b22b69a

all the other changes in this PR are purely changes to support the above code changes. So package name changes, files being moved, code being cleaned up using the new methods and other generic fixes. I've also fixed up all of your @deprecated calls to `getBlock()` methods. Less errors 👍 

I'll be sure to make the commits smaller next time :P

MichaelHillcox commented:
@Direwolf20-MC would it be better for you if I wipe this PR and implement some of the better / more relevant changes in a slower manner allowing you to keep up with what's being changed?

Direwolf20-MC commented:
I appreciate the offer, but I think that'd be a lot of work on you. I'll just go ahead and accept the PR :).

Direwolf20-MC commented:
Thanks I appreciate it! As soon as I find time to accept it I'll get it pulled in, plus I have a few bug fixes of my own to work on once I do. Thanks bud! :)

MajorTuvok commented:
Sorry @Direwolf20 that we overlooked the null ItemStack's...
But thanks @Phylogeny who noticed it :)

MichaelHillcox commented:
I think the `canUse` method logic issue was a bigger issue haha :P

Direwolf20-MC commented:
Lol :) I loaded up the game and none of the gadgets worked at all. It was an easy mistake but definitely made me nervous for a moment! :)

MichaelHillcox commented:
Haha, I bet, I must have misread the original logic or just not read it through. I was testing in creative ( stupid thing to do ) so it was always returning true as the first check is if the user is in creative
